### PR TITLE
[LLVM][runtimes] Prepopulate `LLVM_BUILTIN_TARGETS` with runtimes values

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:
We create the builtins separately because these are required to set up
before others are built. It's configured with a default value if not
specified, but this doesn't respect the runtimes from other targets.
This patch changes the behavior to prepopulate the builtins list with
all the targets that have `compiler-rt` enabled if not overridden by the
user.

### CODE REVIEW

File: llvm/runtimes/CMakeLists.txt [hunk 1]
```diff
@@ -138,37 +142,43 @@ endfunction()
...
+    foreach(name ${LLVM_RUNTIME_TARGETS})
+      if("compiler-rt" IN_LIST RUNTIMES_${name}_LLVM_ENABLE_RUNTIMES)
+        list(APPEND LLVM_BUILTIN_TARGETS ${name})
+      endif()
+    endforeach()
+  endif()
+  if("default" IN_LIST LLVM_BUILTIN_TARGETS)
     builtin_default_target(${compiler_rt_path}
       DEPENDS clang-resource-headers)
+    list(REMOVE_ITEM LLVM_BUILTIN_TARGETS "default")
```
Comment (petrhosek) [inline llvm/runtimes/CMakeLists.txt:159]: What's the reason for doing this? We shouldn't be modifying global variables controlled by the user.

Comment (jhuber6) [reply llvm/runtimes/CMakeLists.txt:159]: That's what the old code does. I could put it in a temporary variable instead if that would be better.

**Comment by jhuber6** [comment]
Also, @petrhosek, what's the specific reason we need to build these in a separate CMake invocation? Doesn't the list of runtimes have an implicit order which we build them? I think we could add a dependency for the `builtins` target within a single CMake invocation if needed. I'm wondering because adding these as separate targets creates a significant overhead to rebuild an extra CMake project.

**Comment by llvm-beanz** [comment]
@jhuber6 the reason I initially implemented builtins and runtimes separately is because the runtime configuration completing successfully and accurately depends on the builtin libraries being available during configuration. We could probably make some educated guess hard-coded assumptions about what the toolchain supports, but you can quickly get off the rails with more exotic target support.

**Comment by jhuber6** [comment]
> @jhuber6 the reason I initially implemented builtins and runtimes separately is because the runtime configuration completing successfully and accurately depends on the builtin libraries being available during configuration. We could probably make some educated guess hard-coded assumptions about what the toolchain supports, but you can quickly get off the rails with more exotic target support.

I see, so it's related to the CMake configure step when it tries to identify features? I can see that being an issue for really specific checks where the compiler defaults to compiler-rt for its runtime instead of some GNU libs. Still unfortunate that it adds such a large overhead to build times.

**Comment by llvm-beanz** [comment]
> I see, so it's related to the CMake configure step when it tries to identify features? I can see that being an issue for really specific checks where the compiler defaults to compiler-rt for its runtime instead of some GNU libs. Still unfortunate that it adds such a large overhead to build times.

To make the point, who has GNU libs at all? I don't in 3 of the 4 configurations I run LLVM's build for.

Yes it is unfortunate, but there are more build configurations than we can reasonably hard code predictions for.

**Comment by jhuber6** [comment]
> > I see, so it's related to the CMake configure step when it tries to identify features? I can see that being an issue for really specific checks where the compiler defaults to compiler-rt for its runtime instead of some GNU libs. Still unfortunate that it adds such a large overhead to build times.
> 
> To make the point, who has GNU libs at all? I don't in 3 of the 4 configurations I run LLVM's build for.
> 
> Yes it is unfortunate, but there are more build configurations than we can reasonably hard code predictions for.

Alright, good to know the motivation behind it. In any case, this patch should fix it to where users will have the builtins built without needing to manually specify an extra variable.

**Comment by jhuber6** [comment]
Ping, this is required for https://github.com/llvm/llvm-project/pull/95304.

**Comment by compnerd** [comment]
I think that @petrhosek and @llvm-beanz definitely should have a look over this - they have been working on the runtimes build, and would have a better idea of how they would like to see it evolve.

File: llvm/runtimes/CMakeLists.txt [hunk 2]
```diff
@@ -138,37 +142,44 @@ endfunction()
 # before the just-built compiler can pass the configuration tests.
 get_compiler_rt_path(compiler_rt_path)
 if(compiler_rt_path)
-  if(NOT LLVM_BUILTIN_TARGETS)
+  # If the user did not specify the targets infer them from the runtimes.
+  set(builtin_targets ${LLVM_BUILTIN_TARGETS})
+  if(NOT builtin_targets)
+    if("compiler-rt" IN_LIST LLVM_ENABLE_RUNTIMES)
+      list(APPEND builtin_targets "default")
+    endif()
```
Comment (compnerd) [inline llvm/runtimes/CMakeLists.txt:150]: I'm not sure that applying DRY here is worth it. I'd rather just duplicate the `builtin_default_target` invocation and remove the string handling for "default".

Comment (jhuber6) [reply llvm/runtimes/CMakeLists.txt:150]: We need the following case to work,
```
   -DLLVM_ENABLE_RUNTIMES=compiler-rt \
   -DRUNTIMES_nvptx64-nvidia-cuda_LLVM_ENABLE_RUNTIMES=compiler-rt               \
   -DRUNTIMES_amdgcn-amd-amdhsa_LLVM_ENABLE_RUNTIMES=compiler-rt                 \
   -DLLVM_RUNTIME_TARGETS="default;amdgcn-amd-amdhsa;nvptx64-nvidia-cuda"
```
Honestly I think we should get rid of the distinction between the "default" target, but that's a much larger refactoring I think @petrhosek wanted to do.

Comment (llvm-beanz) [reply llvm/runtimes/CMakeLists.txt:150]: The "default" target is also important for how the compiler-rt parts of this are built on Darwin, which is a mess.

Comment (petrhosek) [reply llvm/runtimes/CMakeLists.txt:150]: > I'm not sure that applying DRY here is worth it. I'd rather just duplicate the `builtin_default_target` invocation and remove the string handling for "default".

I have a WIP change that removes the `"default"` altogether since that part is a relic, but it needs some more testing to make sure I don't break any of existing configurations.

**Comment by jhuber6** [comment]
ping

**Comment by jhuber6** [comment]
@petrhosek

File: llvm/runtimes/CMakeLists.txt [hunk 3]
```diff
@@ -17,7 +17,11 @@ foreach(proj ${LLVM_ENABLE_RUNTIMES})
 endforeach()
 
 function(get_compiler_rt_path path)
-  foreach(entry ${runtimes})
+  set(all_runtimes ${runtimes})
+  foreach(name ${LLVM_RUNTIME_TARGETS})
+    list(APPEND all_runtimes RUNTIMES_${name}_LLVM_ENABLE_RUNTIMES)
+  endforeach()
+  foreach(entry ${all_runtimes})
```
Comment (petrhosek) [inline llvm/runtimes/CMakeLists.txt:24]: This looks like an unrelated change that could land separately.

**Review by petrhosek — APPROVED** [review]

File: llvm/runtimes/CMakeLists.txt [hunk 4]
```diff
@@ -17,7 +17,11 @@ foreach(proj ${LLVM_ENABLE_RUNTIMES})
 endforeach()
 
 function(get_compiler_rt_path path)
-  foreach(entry ${runtimes})
+  set(all_runtimes ${runtimes})
+  foreach(name ${LLVM_RUNTIME_TARGETS})
+    list(APPEND all_runtimes RUNTIMES_${name}_LLVM_ENABLE_RUNTIMES)
```
Comment (petrhosek) [inline llvm/runtimes/CMakeLists.txt:22]: I think you need to evaluate the variable otherwise you're just going to add `RUNTIMES_${name}_LLVM_ENABLE_RUNTIMES` as an element to the list.
```suggestion
    list(APPEND all_runtimes ${RUNTIMES_${name}_LLVM_ENABLE_RUNTIMES})
```

File: llvm/runtimes/CMakeLists.txt [hunk 5]
```diff
@@ -17,7 +17,11 @@ foreach(proj ${LLVM_ENABLE_RUNTIMES})
...
+  endforeach()
```
Comment (petrhosek) [inline llvm/runtimes/CMakeLists.txt:23]: You should consider deduplicating the items in the list here:
```suggestion
  endforeach()
  list(REMOVE_DUPLICATES all_runtimes)
```
